packages/relayer/test/Relayer.test.ts

```diff
@@ -178,28 +222,132 @@ describe('Relayer', () => {
       const value = amount
 
-      it('deposits the amount correctly', async () => {
-        const previousSmartVaultBalance = await relayer.getSmartVaultBalance(smartVault.address)
+      context('when the used quota is zero', () => {
+        it('deposits the amount correctly', async () => {
+          const previousSmartVaultBalance = await relayer.getSmartVaultBalance(smartVault.address)
 
-        await relayer.deposit(smartVault.address, amount, { value })
+          await relayer.deposit(smartVault.address, amount, { value })
 
-        const currentSmartVaultBalance = await relayer.getSmartVaultBalance(smartVault.address)
-        expect(currentSmartVaultBalance).to.be.equal(previousSmartVaultBalance.add(amount))
-      })
+          const currentSmartVaultBalance = await relayer.getSmartVaultBalance(smartVault.address)
+          expect(currentSmartVaultBalance).to.be.equal(previousSmartVaultBalance.add(amount))
+        })
 
-      it('increments the relayer balance properly', async () => {
-        const previousRelayerBalance = await ethers.provider.getBalance(relayer.address)
+        it('increments the relayer balance properly', async () => {
+          const previousRelayerBalance = await ethers.provider.getBalance(relayer.address)
 
-        await relayer.deposit(smartVault.address, amount, { value })
+          await relayer.deposit(smartVault.address, amount, { value })
 
-        const currentRelayerBalance = await ethers.provider.getBalance(relayer.address)
-        expect(currentRelayerBalance).to.be.equal(previousRelayerBalance.add(amount))
+          const currentRelayerBalance = await ethers.provider.getBalance(relayer.address)
+          expect(currentRelayerBalance).to.be.equal(previousRelayerBalance.add(amount))
+        })
+
+        it('emits an event', async () => {
+          const tx = await relayer.deposit(smartVault.address, amount, { value })
+
+          await assertEvent(tx, 'Deposited', { smartVault, amount })
+        })
       })
 
-      it('emits an event', async () => {
-        const tx = await relayer.deposit(smartVault.address, amount, { value })
+      context('when the used quota is not zero', () => {
+        let authorizer: Contract, smartVault: Contract, smartVaultOwner: SignerWithAddress
+        let usedQuota: BigNumber, currentQuota: BigNumber, paidQuota: BigNumber
+        let amount: BigNumber, value: BigNumber, toDeposit: BigNumber
+
+        beforeEach('deploy smart vault', async () => {
+          smartVaultOwner = await getSigner()
+          // eslint-disable-next-line prettier/prettier
+          ;({ authorizer, smartVault } = await deployEnvironment(smartVaultOwner))
+        })
+
+        beforeEach('set sender', async () => {
+          relayer = relayer.connect(owner)
+        })
+
+        beforeEach('set maximum quota', async () => {
+          const maxQuota = fp(10000)
+          await relayer.setSmartVaultMaxQuota(smartVault.address, maxQuota)
+        })
```

Comment:
we shouldn't use the owner for everything here:

```suggestion
        beforeEach('set maximum quota', async () => {
          const maxQuota = fp(10000)
          await relayer.connect(owner).setSmartVaultMaxQuota(smartVault.address, maxQuota)
        })
```